feat: Split mrack into multiple rpms

Tiboris opened the pull request:
Splitting mrack to:
    - mrack
    - mrack-cli
    - python3-mracklib
    - python3-mrack-aws
    - python3-mrack-beaker
    - python3-mrack-openstack
    - python3-mrack-podman
    - python3-mrack-virt
   
    where python3-mracklib contains only static provider and all the
    scripts and provisioning base logic.
    we can install mrack-{aws,beaker,openstack,podman,virt}
    providers in addition to static provider extending
    the functionality and create modular solution where only
    needed providers can be installed.
    Package mrack contains all supported providers.
    Package mrack-cli contains mrack command.
    We are using Suggests directives for OpenStack dependencies
    so in fedora where these are not available build will pass.

Resolves: https://github.com/neoave/mrack/issues/113 

This MR depends on:
- [x] https://github.com/neoave/mrack/pull/191 
- [x] https://github.com/neoave/mrack/pull/194 
- [x] discuss the pacakge division

pvoborni commented on mrack.spec:
```diff
@@ -1,5 +1,5 @@
 # Created by pyp2rpm-3.3.5
-%global srcname mrack
```
typo: `pyton3`

pvoborni replied:
And actually, python applications doesn't have to start with `python3`. 
As https://docs.fedoraproject.org/en-US/packaging-guidelines/Python/#_application_naming says:

> Packages that primarily provide applications, services or any kind of executables SHOULD be named according to the general [Fedora naming guidelines](https://docs.fedoraproject.org/en-US/packaging-guidelines/Naming/) (e.g. ansible).

Probably the most correct thing would be to have packages:

- mrack (packaging mrack CLI)
- python3-mrack (core of the mrack - the library to be imported e.g. in tmt)
- python3-mrack-openstack (the provider)

Tiboris replied:
I have re-pushed without the patch adding the python- prefix.
We should decide how to divide the package itself i have pushed this PR to initiate discussion basically as we need to agree on something which makes sense for us as well.

My question is, should we create package per each provider from (AWS, Virt, Beaker, Openstack, Podman) and keep static as a base for the library or should we make 

```
└─ mrack   
       └─  python3-mrack-all (maybe without this part)
                     └─ python3-mrack-aws
                     │          └─  python3-mrack (Static)  
                     ├─ python3-mrack-beaker
                     │          └─  python3-mrack (Static)  
                     ├─ python3-mrack-openstack
                     │          └─  python3-mrack (Static)  
                     ├─ python3-mrack-podman
                     │          └─  python3-mrack (Static)  
                     └─ python3-mrack-virt
                                └─  python3-mrack (Static)  

```

Tiboris replied:
How would you divide the package?

pvoborni reviewed (commented):
I didn't test, so these are just my initial comments after reading the PR. It mostly looks good to me.

Tiboris commented:
@praiskup Would you mind having a look here?

praiskup commented on mrack.spec:
```diff
@@ -12,48 +12,120 @@ Source0:        https://github.com/neoave/mrack/releases/download/v%{version}/mr
+Requires:       %{name}-core
+Requires:       %{name}-aws
+Requires:       %{name}-beaker
+Suggests:       %{name}-openstack
```
This isn't needed IMO, the `mrack-openstack` sub-package is installable because *that* package uses "only" Suggests for problematic deps.

Tiboris replied:
So at this level `Requires` is safe correct?

praiskup commented on mrack.spec:
```diff
@@ -1,7 +1,7 @@
 %global srcname mrack
 
-Name:           %{srcname}
+Name:           mrack
```
A nit ... you could even drop the %srcname definition now?  I don't think it was necessary to move from `srcname` to just `name`, but admittedly the spec looks prettier now.

Tiboris replied:
Thanks! Yeah I forgot to remove the `srcname`

praiskup commented:
So previously there was just `mrack`, and also the `mrack` package stays here ... serving the same purpose as before, bringing in all the new deps.  So you don't have to do any artificial dependency "obsoletes" hacks or so, good (you just update to an updated mrack.rpm, and all the deps will be installed automatically).  Overall it looks fine to me.

Tiboris commented:
@pvoborni do we want the mrack-cli package? it would contain mrack script only

Tiboris commented:
Right now the packages should reflect this dependency graph:
```
└─             mrack   
                     └─  mrack-core (Static)    
                     └─ mrack-aws
                     │     └─  mrack-core (Static)    
                     ├─ mrack-beaker
                     │     └─  mrack-core (Static)    
                     ├─ mrack-openstack
                     │      └─  mrack-core (Static)   
                     ├─ mrack-podman
                     │      └─  mrack-core (Static)   
                     └─mrack-virt
                               └─  mrack-core (Static)  

```

praiskup commented:
> mrack-all (virtual package which requires all providers in this repo packaged for given OS)

This sounds doable?  Not the most common and natural pattern, you e.g. need to take an explicit care of
the upgrade path (Obsoletes etc.).  But overall looks like OK to me as well (IMVHO a matter of taste).

Tiboris commented:
@praiskup does it make sense to have mrack-core as a dependency of mrack if it is already transitive of each dependent package?

praiskup commented:
>  does it make sense to have mrack-core as a dependency of mrack

@Tiboris technically it is not necessarily, when the main (mrack or mrack-all) package is only supposed to be a "metapackage".   If not a metapage (== contains at least some files that depend on core) it is better to stay explicit (even though the dependency could already exist implicitly, generated by rpmbuild itself ... I'd have to test).

praiskup commented:
> So what happened to the mrack-all and python3- prefix?

+1 for python3- prefixes

Tiboris commented:
@pvoborni I have fixed the python prefix and updated spec file to reflect things we agreed on in our discussion, shortly:

Splitting mrack to:
- mrack  # all the supported providers and cli
- mrack-cli # mrack command with static provider
- python3-mracklib # core mrack libraries with static provider
- python3-mrack-aws 
- python3-mrack-beaker
- python3-mrack-openstack
- python3-mrack-podman
- python3-mrack-virt

@praiskup I hope that the python prefix is fine now I did not find any directive for such prefix thus i hard-coded it did i tried poorly (to find the directive)?

Tiboris commented:
Also I have updated `README.md` to suggest installation steps when only a subset of functionality is required.

Tiboris commented:
PR is ready for review.

praiskup commented on mrack.spec:
```diff
@@ -1,59 +1,141 @@
+Requires:       python3-%{name}-beaker == %{version}-%{release}
+Requires:       python3-%{name}-openstack == %{version}-%{release}
+Requires:       python3-%{name}-podman == %{version}-%{release}
+Requires:       python3-%{name}-virt == %{version}-%{release}
```
Requires need to be specified before the %description tag (otherwise they are just part of %description).

Tiboris replied:
Fixed

praiskup commented:
> @praiskup I hope that the python prefix is fine now I did not find any directive for such prefix thus i hard-coded it did i tried poorly (to find the directive)?

No, seems fine -  python3- prefix is usually hard-coded in every %package, %description, %files tag....

pvoborni commented on README.md:
```diff
@@ -121,9 +121,36 @@ pip install mrack
+---
+**NOTE**
+
+Above command will install `mrack` package which depends (contains) following packages:
```
`depends` -> `depends on` ?

Tiboris replied:
Let me fix this

pvoborni reviewed (approved):
The changes looks good to me - all the names and deps. 

But I did not test it, thus might have missed something.

praiskup commented on mrack.spec:
```diff
@@ -1,59 +1,145 @@
+%{python3_sitelib}/%{name}-%{version}-py%{python3_version}.egg-info
+%exclude %{python3_sitelib}/%{name}/providers/utils/{,__pycache__/}osapi.py
+%exclude %{python3_sitelib}/%{name}/providers/{,__pycache__/}{%{provider_plugins}}.*
+%exclude %{python3_sitelib}/%{name}/transformers/{,__pycache__/}{%{provider_plugins}}.*
```
I'd personally do `%dir %{python3_sitelib}/%{name}` and then drop the %exclude statements

Tiboris replied:
As discussed this would create a long list of includes doing it this way... Lets keep the excludes only ?

praiskup replied:
taking back this one, the list of "include" statements would be too large...

praiskup commented on mrack.spec:
```diff
@@ -1,59 +1,145 @@
+Requires:       python3-%{name}-podman == %{version}-%{release}
+Requires:       python3-%{name}-virt == %{version}-%{release}
+
+%global __requires_exclude asyncopenstackclient
```
Please document why you are doing this :-)

Tiboris replied:
Updated

praiskup commented:
looks fine to me
